merge getReportIcons and getAvatarSources functions

marcochavezf opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->
cc @TomatoToaster @roryabraham 

### Details
<!-- Explanation of the change or anything fishy that is going on -->
This PR merges the `getAvatarSources` into the `getReportIcons` function, and it's called when an option row is created (in LHN, Search bar, and Invite New members). This reduces complexity in the code to handle report icons and updates the workspace avatar for all option rows when it's updated from the Workspace page.

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/Expensify/issues/199856

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Login as admin of a workspace (if you don't have one follow the test steps of [PR](https://github.com/Expensify/App/pull/7852#issue-1146340425))
2. Change the workspace avatar from the Workspace page and verify it's updated in all the places where the workspace avatar is shown.
3. Additionally, verify the avatars are displayed correctly in the Search bar and in the Invite new members page.

- [x] Verify that no errors appear in the JS console

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.


<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
4. Upload an image via copy paste
5. Verify a modal appears displaying a preview of that image
--->
Follow the same steps from `Test` section.

- [ ] Verify that no errors appear in the JS console

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->
https://user-images.githubusercontent.com/6829422/161865849-0fb3cdaa-057d-4b52-ab23-520a21dc930d.mov

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->
<img width="325" alt="Screen Shot 2022-04-05 at 18 36 00" src="https://user-images.githubusercontent.com/6829422/161868022-d56cdc46-ee5c-45f5-905d-bb01273f207b.png">

<img width="325" alt="Screen Shot 2022-04-05 at 18 51 12" src="https://user-images.githubusercontent.com/6829422/161869373-74f71790-aee7-4608-bed3-cd0e07ff545a.png">

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->
https://user-images.githubusercontent.com/6829422/161868967-a89d5cf7-cb27-4338-854e-53fd8fca7188.mov

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->
https://user-images.githubusercontent.com/6829422/161867880-1eaa0c27-263e-48c9-afd1-cbc5dcb70b19.mov

#### Android
<!-- Insert screenshots of your changes on the Android platform-->
https://user-images.githubusercontent.com/6829422/161868817-6166ea50-e4ee-4115-aee7-5598e97e2552.mov

marcochavezf committed:
avoid storing report icons into Onyx

marcochavezf committed:
get avatar source directly from report.icons

marcochavezf committed:
remove getReportIcons from PersonalDetails

marcochavezf committed:
use getReportIcons in getAvatarSources

marcochavezf committed:
get personalDetails from Onyx to get avatar urls

marcochavezf committed:
merge getReportIcons and getAvatarSources into one function

marcochavezf committed:
fix single avatars by returning an array

marcochavezf committed:
fix default icon component if avatar is an empty string

marcochavezf committed:
add comment for group chats

marcochavezf committed:
return empty string for firstName in sortedParticipants

marcochavezf committed:
change fn name getAvatarSources to getReportIcons

marcochavezf committed:
add defaultIcon param to getReportIcons

marcochavezf commented:
>I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
The functionality to update or remove the workspace avatar only works online, but this functionality will need to change once the offline first procedures are completed.

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -78,6 +78,12 @@ Onyx.connect({
+Onyx.connect({
+    key: ONYXKEYS.PERSONAL_DETAILS,
+    callback: val => formattedPersonalDetails = val,
+});
```
**thought:** we were avoiding this subscription by passing in the `personalDetails` whenever we use `getOptions()`. 

**question:** why do we need this now?

marcochavezf replied:
This is used to get the avatars for group chats, an alternative solution is to pass the `personalDetails` using `withOnyx` to the components using `getReportIcons`.

marcochavezf replied:
Update: I noticed `personalDetails` is already passed (using `withOnyx`) in a few components using `getReportIcons`, the only component that is not passing it is `ReportActionItemCreated`. I will remove the subscription and pass `personalDetails` to this component.

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -206,33 +212,53 @@ function hasReportDraftComment(report) {
-        }
-        if (ReportUtils.isPolicyExpenseChat(report)) {
-            return Expensicons.Workspace;
+function getReportIcons(report, defaultIcon = null) {
```
Unsure how hard this would be to change, but this feels like more of a `ReportUtils` responsibility maybe?

Looking at what `OptionsListUtils` is doing and it's unclear whether this method belongs here 🤔

marcochavezf replied:
Fair point, yeah I leave it there because it was always in that file, but makes more sense to move it to `ReportUtils`

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -814,47 +835,10 @@ function getCurrencyListForSections(currencyOptions, searchValue) {
- * @param {Object} personalDetails
- * @returns {Array<String>}
- */
-function getReportIcons(report, personalDetails) {
```
Nice consolidation of this logic.

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -1506,11 +1505,6 @@ function createPolicyRoom(policyID, reportName, visibility) {
-            // Make sure the report has its icons set
-            const report = allReports[reportID];
-            const icons = OptionsListUtils.getReportIcons(report, {});
-            Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT}${reportID}`, {icons});
```
**clarification:** we are not saving the icons because we build them from personal details now? Or something else?

tgolen commented on src/pages/home/HeaderView.js:
```diff
@@ -101,7 +101,7 @@ const HeaderView = (props) => {
     const avatarTooltip = isChatRoom ? undefined : _.pluck(displayNamesWithTooltips, 'tooltip');
     const shouldShowSubscript = isPolicyExpenseChat && !props.report.isOwnPolicyExpenseChat;
-    const avatarIcons = OptionsListUtils.getAvatarSources(props.report);
+    const avatarIcons = OptionsListUtils.getReportIcons(props.report);
```
```suggestion
    const reportIcons = OptionsListUtils.getReportIcons(props.report);
```

tgolen commented on src/pages/home/report/ReportActionItemCreated.js:
```diff
@@ -25,7 +25,7 @@ const defaultProps = {
 const ReportActionItemCreated = (props) => {
     const isPolicyExpenseChat = ReportUtils.isPolicyExpenseChat(props.report);
-    const avatarIcons = OptionsListUtils.getAvatarSources(props.report);
+    const avatarIcons = OptionsListUtils.getReportIcons(props.report);
```
```suggestion
    const reportIcons = OptionsListUtils.getReportIcons(props.report);
```

tgolen commented on src/pages/ReportDetailsPage.js:
```diff
@@ -133,7 +133,7 @@ class ReportDetailsPage extends Component {
                             <View style={styles.mb4}>
                                 <RoomHeaderAvatars
-                                    avatarIcons={OptionsListUtils.getAvatarSources(this.props.report)}
+                                    avatarIcons={OptionsListUtils.getReportIcons(this.props.report)}
```
```suggestion
                                    reportIcons={OptionsListUtils.getReportIcons(this.props.report)}
```

marcochavezf replied:
Ah, this suggestion changes the property name of the`RoomHeaderAvatars` component, and we're already using the property name `avatarIcons` consistently in other Avatar similar components. Is ok if I change `avatarIcons` to `reportIcons` to make it consistent on all the similar components? Or maybe just to `icons`?

tgolen replied:
Yeah, there would definitely be some cascading effects for renaming this. I would be very happy to call everything `icons` and keep it consistent across all components that it touches.

marcochavezf committed:
remove personalDetails subscription and add it as param in getReportIcons

marcochavezf committed:
pass personalDetails for getReportIcons for ReportActionItemCreated

marcochavezf committed:
add personalDetails param to getReportIcons usages

marcochavezf committed:
move personalDetails to createOption to pass it to getReportIcons

marcochavezf committed:
move getReportIcons to reportUtils

marcochavezf committed:
fix errors in getReportIcons

marcochavezf committed:
move getDefaultAvatar to reportUtils

marcochavezf committed:
fix usages of getDefaultAvatar

marcochavezf committed:
fix getDefaultAvatar in ProfilePage

marcochavezf committed:
fix usages of getReportIcons

marcochavezf committed:
remove unnecessary jsdoc params

marcochavezf committed:
add policies as param to getReportIcons

marcochavezf committed:
include policies prop via withOnyx to ReportActionItemCreated

marcochavezf committed:
pass policies in when getReportIcons is called

marcochavezf committed:
change const name avatarIcons to reportIcons

marcochavezf committed:
fix issue when personalDetail is not defined

marcochavezf committed:
Merge branch 'main' into marco-mergeAvatarIconFunctions

marcochavezf commented:
Update: Sorry if that explanation of my changes will sound a bit hairy, but I think the code now looks cleaner and brought one benefit that I'm going to explain below:

- I removed the subscription to `personalDetails` that I had added to `OptionListUtils` (which was required for the new `getReportICons` function), but that removal also required to pass `personalDetails` to `createOption`, so I updated `createOption` params to receive `personalDetails`.  Also I added the `logins` param to create `personalDetailList` inside of `createOption` (which removes a bit of DRY code for `personalDetailList`).
- I moved the `getReportICons` to `ReportUtils` and that required to add the parameter `policies` to `getReportICons` (in `OptionListUtils` we have a subscription to `policies`). 
- In `ReportActionItemCreated`  now we require to pass `policies` via `withOnyx` to create the icons. This change is good because I noticed we were not updating the workspace avatar in the welcome message of the workspace chat, and now it's fixed gracefully with the new version of `getReportICons` which needs to receive the`policies` as param (long live pure functions!):

https://user-images.githubusercontent.com/6829422/162323920-396784f4-4bb2-44de-a7a6-456d617b2d68.mov

The only pending suggestion is about changing `avatarIcons` to `reportIcons` (or `icons`) which is discussed [here](https://github.com/Expensify/App/pull/8499#discussion_r845438054). Let me know your thoughts if sounds good to include that change in this PR, besides that, the changes are ready to be reviewed again.

marcochavezf committed:
rename avatarIcons to icons in MultipleAvatars

marcochavezf committed:
change avatarIcons to icons in RoomHeaderAvatars

marcochavezf committed:
update usages of RoomHeaderAvatars

marcochavezf committed:
change getReportIcons to getIcons

marcochavezf commented:
Update: I renamed the properties `avatarIcons` to `icons` in the related Avatar components, also I renamed the function `ReportUtils.getReportIcons` to `ReportUtils.getIcons` which I think looks cleaner and less redundant because the first parameter is a `report` object.

tgolen reviewed (approved):
Ah, that looks really good now. Thanks for taking the time to update those!

TomatoToaster commented:
I think we'd be good to merge this but want to double check @marcaaron because you're also tagged for review.

marcaaron reviewed (approved):
Super nice changes thanks @marcochavezf

marcochavezf commented:
Thank you all!

marcochavezf commented:
The PR already passed the tests, removing the `Emergency` label.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2197582315) to staging by @marcochavezf in version: 1.1.56-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2223133075) to production by @francoisl in version: 1.1.56-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
